File: Java/798.java
__________________________________________________________________________________________________
sample 2 ms submission
class Solution {
    public int bestRotation(int[] A) {
        int n = A.length;
        int[] score = new int[n + 1];
        for (int i = 0; i < n; i++) {
            if (A[i] > i) {
                score[i + 1]++;
                score[n - (A[i] - i) + 1]--;
            } else {
                score[0]++;
                score[i - A[i] + 1]--;
                score[n - (n - i) + 1]++;
                score[n]--;
            }
        }
        int res = 0;
        int max = score[0];
        for (int i = 1; i < n; i++) {
            score[i] += score[i - 1];
            if (score[i] > max) {
                max = score[i];
                res = i;
            }
        }
        return res;
    }
}
__________________________________________________________________________________________________
sample 39336 kb submission
class Solution {
    public int bestRotation(int[] A) {
        int n = A.length;
        int[] rotateLost = new int[n];   
        for(int i = 0; i<n; i++)
            rotateLost[(i-A[i]+n+1)%n]++;
        int[] score = new int[n];
        int max = score[0];
        int result = 0;
        for(int i = 1; i<n; i++){
            score[i] = score[i-1]-rotateLost[i]+1;
            if(max<score[i]){
                max = score[i];
                result = i;
            }
        }
        
        return result;
    }
}
__________________________________________________________________________________________________

File: Python3/798.py
__________________________________________________________________________________________________
sample 196 ms submission
# [4 1 3 2 0]
class Solution(object):
    def bestRotation(self, A):
        n = len(A)
        out = [1] * n

        for i in range(n):
            out[(i - A[i] + 1) % n] -= 1

        for i in range(1, n):
            out[i] += out[i-1]

        return out.index(max(out))
__________________________________________________________________________________________________
sample 14972 kb submission
class Solution:
    def bestRotation(self, A: List[int]) -> int:
        n = len(A)
        accu = [0] * (n+1)
        for i, a in enumerate(A):
            if a >= n:
                continue
                
            if i <= a:
                # a <= i+K <= n-1
                accu[a-i] += 1
                accu[n-i] -= 1
            else:
                # 
                accu[0] += 1
                accu[n-i] -= 1
                #  i + K >= n + a
                accu[n+a-i] += 1
                accu[n] -= 1
 
        realk = 0 # realk if left rotate, k is right rotate
        maxsums = 0
        sums = 0
        for k, a in enumerate(accu[:-1]):
            sums += a
            if sums > maxsums or (sums == maxsums and realk != 0):
                realk = (n - k ) % n
                maxsums = sums
     
        return realk
__________________________________________________________________________________________________
